proposals/0000-warning-pragmas-with-categories.rst

```diff
@@ -0,0 +1,252 @@
+proposal to introduce::
+
+    {-# WARNING [head,partial] head "This is a partial function, it throws an error on empty lists." #-}
+    {-# WARNING [tail,partial] tail "This is a partial function, it throws an error on empty lists." #-}
+
+By default, ``-Wwarnings-deprecations`` is enabled, so the appropriate warning
+would be displayed at any use of ``head`` or ``tail``.  However, users may
+control this behaviour to their liking:
+
+* Users who prefer to use the functions unimpeded may specify
+  ``-Wno-warnings:partial`` to suppress all warnings annotated with the
+  ``partial`` warning category, while still seeing other warnings.
+
+* Users who do not wish to use the functions at all may specify
+  ``-Werror=warnings:partial`` so that any use of the functions will be reported
+  as an error, while other warnings are unaffected.
+
+* Since multiple warning categories can be applied to an identifier, users may
+  distinguish more finely if needed, for example ``-Werror -Wwarn=warnings:head
+  -Wno-warnings:tail`` could be used in a codebase where most warnings are
+  considered errors, except uses of ``head`` are merely warned about, and uses
+  of ``tail`` are not mentioned.
+
+
+Proposed Change Specification
+-----------------------------
+
+The `WARNING and DEPRECATED pragmas
+<https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/exts/pragmas.html#warning-deprecated-pragma>`_
+may be immediately followed by a comma-separated list of warning categories in
+square brackets. Each warning category is an identifier beginning with a letter
+and followed by valid identifier characters or dashes.
+
+More formally, the grammar of declarations is extended as follows:
+
+============  =  =====================================================================
+*decl*        →  ``{-# WARNING`` [*categories*] *things* *strings* ``#-}``
+*decl*        →  ``{-# DEPRECATED`` [*categories*] *things* *strings* ``#-}``
+
+*categories*  →  ``[`` *category1*, ..., *categoryN* ``]``
+*category*    →  (*small* | *large*) { *small* | *large* | *digit* | ``'`` | ``-`` }
+
+*things*      →  *thing1*, ..., *thingN*
+*thing*       →  *varid* | *conid*
+*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
+============  =  =====================================================================
+
+The list of categories can be omitted entirely, so this subsumes the existing
+syntax for ``WARNING`` and ``DEPRECATED`` pragmas.  That is, [*categories*]
+means that the presence of the *categories* non-terminal is optional; if present
+it is a list in square brackets (hence the use of typewriter font ``[``
+... ``]`` in the definition of the *categories* production).
+
+The *category* non-terminal subsumes both *varid* and *conid*, so it is always
+possible to use the name of the thing to which a warning is being attached as
+one of the categories.  The dash character (``-``) is permitted as a character
```

Comment:
Neat idea, but do we really want tools that deal with GHC command lines deal with all that complexity? Unicode emojis etc?

Also, a name can be many things. Can I use fully qualified names? Maybe even package qualified names?

Also, multiple categories is something we don't have yet. Does that mean it matters in which order I pass `-Werror=` or `-Wno-` etc. 

I'd aim for a single category only consisting of lower case letters, dashes, and maybe numbers, and keep it simple.

Reply:
> Neat idea, but do we really want tools that deal with GHC command lines deal with all that complexity? Unicode emojis etc?

Fair question. I don't have much insight into the details of cross-platform CLI handling, so I don't know what an appropriate set of restrictions on characters would be. (Obviously one could permit only ASCII lowercase letters, digits, and dashes, but might that be an annoying for restriction for users in some locales?)

One nice aspect of the present proposal is that every Haskell identifier name is a valid warning category, so library authors can choose to have the category name match the identifier name. That's also why I suggested allowing categories to start with uppercase letters. But I don't hugely care about the exact choices here.

> Also, a name can be many things. Can I use fully qualified names? Maybe even package qualified names?

No, this is mentioned briefly in the Costs and Drawbacks section.

> Also, multiple categories is something we don't have yet. Does that mean it matters in which order I pass -Werror= or -Wno- etc. (Answer below, I know, but I wish I would not have to ask :-))

Perhaps we don't directly have multiple categories, but we do have warning groups (e.g. `-W`, `-Wcompat`, `-Wall`) which amount to much the same thing, just with a restricted list of groups. In particular, we already have order-dependence (e.g. `-Wno-missing-signatures -Wall` is equivalent to `-Wall` and not `-Wall -Wno-missing-signatures`).

I think the "tagging" approach that multiple categories permits is a simple way to get behaviour like the existing warning groups, without needing a predefined taxonomy.